pep-0722.rst

```diff
@@ -0,0 +1,271 @@
+``pip-run`` tools, so it is simply formalising existing behaviour, rather than
+defining a brand new capability.
+
+
+Rationale
+=========
+
+In general, the Python packaging ecosystem is focused on "projects", which are
+structured around a directory containing your code, some metadata and any
+supporting files such as tests, automation, etc. However, beginners (and quite a
+few more experienced developers) will often *start* a project by just opening up
+an editor and writing a simple script. In some cases, the project might outgrow
+the "simple script" phase and get restructured into a more conventional project
+directory, but not all do, and even those that do still need to be usable in the
+"simple script" phase.
+
+These days, the idea that "simple scripts can just use the stdlib" is becoming
+less and less practical, as the quantity, quality and usability of libraries on
+PyPI steadily increases. For example, using the stdandard ``urllib`` library
+rather than something like ``requests`` or ``httpx`` makes correct code
+significantly harder to write.
+
+Having to consider "uses 3rd party libraries" as the break point for moving to a
+"full scale project" is impractical, so this PEP is designed to allow a project
+to use external libraries while still remaining as a simple, standalone script.
+
+Of course, *declaring* your dependencies isn't sufficient by itself. You need to
+install them (probably in some sort of virtual environment) so that they are
+available when the script runs. This PEP does not cover environment management,
+as tools like `pip-run <https://pypi.org/project/pip-run/>`__ and `pipx
+<https://pypi.org/project/pipx/>`__ already offer that ability. But by
+standardising the means of declaring dependencies, this PEP allows scripts to
+remain tool-independent.
+
+
+Specification
+=============
+
+Any Python script may contain a *dependency block*, which is a specially
+structured comment block. The format of the dependency block is as follows:
+
+* A single comment line containing the (case sensitive) text "Requirements:"
+* A series of comment lines containing :pep:`508` requirement specifiers.
+* An empty comment or blank line.
+
+To be recognised as a "comment line", the line must start with a ``#`` symbol.
+Leading and trailing whitespace on the comment line is ignored.
+
+The dependency block may occur anywhere in the file. There MUST only be a single
+dependency block in the file - tools consuming dependency data MAY simply
+process the first dependency block found. This avoids the need for tools to
+process more data than is necessary. Stricter tools MAY, however, fail with an
+error if multiple dependency blocks are present.
+
+Example
+-------
+
+The following is an example of a script with an embedded dependency block::
+
+    #!/usr/bin/env python
+
+    # In order to run, this script needs the following 3rd party libraries
+    #
+    # Requirements:
+    #    requests
```

Comment:
Perhaps these requirements should follow a similar declarative syntax to pyproject.toml, specifically to allow for versioning? Accepting a dependency, without hinting as to the potential required version makes ongoing maintenance more difficult, IMHO.

Reply:
Discussion should happen on the Python Discourse thread noted in the PEP header. Reviews here should *only* be for typos or procedural matters. Please raise your point on Discourse.

Unless you just mean you'd like me to include a version spec in the example, something like `requests>=2.30`? I assumed that the statement that PEP 508 requirements were allowed made that clear, and I prefer having realistic examples. But if there's wider support for doing that, I can - I don't intend to cover all of PEP 508, though...